js/plugins/vertexai/src/gemini.ts

```diff
@@ -532,8 +532,19 @@ export function geminiModel(name: string, vertex: VertexAI): ModelAction {
         };
       } else {
         let result: GenerateContentResult | undefined;
+        let chat = client.startChat(chatRequest);
         try {
-          result = await client.startChat(chatRequest).sendMessage(msg.parts);
+          const toolResponses = msg.parts.filter(
+            (part) => part.functionResponse
+          );
+          const nonToolParts = msg.parts.filter(
+            (part) => !part.functionResponse
+          );
+          // The Gemini API does not support sending tool responses in the same message as other parts
+          for (const part of toolResponses) {
+            chat.sendMessage([part]);
+          }
```

Comment:
what's going on here? it looks like it makes the chat call for each tool response separately, but not doing anything with the response?

Reply:
I wasn't 100% on this, but the thinking was the chat instance in the sdk will keep hold of the responses, but I need to look at this again. For instance surely it needs to be awaited if it works how I thought it did

Reply:
`sendMessage` makes a call to the API... and returns the response. This code will result in multiple LLM calls instead of 1, which sounds.... expensive. :)

Reply:
Ah we should definitely group them into one call haha. The issue in question is that a FunctionResponse isn't allowed to be in the same message as other parts, I believe

Reply:
I think it's just a matter of splitting them into separate messages... and moving the non tool parts into history.